### PR TITLE
Extract Lua engine into a separate package

### PR DESCRIPTION
**Pull request by tarleb**

### CODE REVIEW

**Comment by tarleb** [comment]
This makes it easy to build pandoc without Lua. The effect on the resulting binary is unexpectedly *huge*, shrinking the (unoptimized) binary by about 40%.

**Comment by jgm** [comment]
I tried this on an M1 mac.  I saw the following difference in stripped binary sizes:
with lua: 154 MB
without lua: 151 MB
Also, in the build with lua, filters and custom writers don't seem to work; it just hangs when I try to use them, and I have to press ^C twice to stop.

Separate thought: maybe a ScriptingEngine should have  a textual description field, which we can include in the `pandoc --version` output (e.g. "compiled with Lua 5.4 scripting support").

Another separate thought: if we have a `lua` flag on pandoc-cli to disable Lua support, we should also have a `server` flag so taht server support can be disabled.

**Comment by tarleb** [comment]
> I tried this on an M1 mac. I saw the following difference in stripped binary sizes:
> with lua: 154 MB
> without lua: 151 MB

That actually helps me sleep much easier. I thought the general idea of "almost free" functionality through Lua was fundamentally flawed. But that means that there is something strange happening on Linux, because here I get
with lua: 194 MB
without lua: 116 MB

The issue of pandoc hanging appears to be related to the mediabag: using `pandoc-lua` works fine unless the script uses `pandoc.mediabag.fetch`. Not sure yet what's going on there.

**Comment by jgm** [comment]
> The issue of pandoc hanging appears to be related to the mediabag: using pandoc-lua works fine unless the script uses pandoc.mediabag.fetch. Not sure yet what's going on there.

I'm not talking about `pandoc-lua`.  I'm talking about using `pandoc` with `--lua-filter` or a custom reader or writer. This hangs (even if you provide the name of a nonexistent filter).

**Comment by tarleb** [comment]
Oh, interesting. I assumed that both bugs are very much related, and fixing one would fix the other. But the fact that it hangs on a nonexistent filter makes me doubt that assumption.

I kind of hope that this is one of those bugs whose fix will improve the whole system. My first thought is "lazyness", but I was only half-right about that the last time.

**Comment by jgm** [comment]
I assume you're not seeing the same thing on linux, then?
Maybe worth adding a few sanity checks to the CI (at least temporarily), so we can see how it works on all architectures/compiler versions.

**Comment by jgm** [comment]
PS. I'll add a `server` flag in the main branch.

**Comment by tarleb** [comment]
Now I'm seeing it on Linux, too. Presumably I tested that with a different version without realizing.

It seems that pandoc hangs in Text.Pandoc.Class.IO.readFileStrict, but not always. I'm still at a loss so far.

**Comment by tarleb** [comment]
Ah, found it. I created a loop by replacing `Text.Pandoc.Class.IO` with `Text.Pandoc.Class` in the definition of PandocLua. :man_facepalming:

**Comment by tarleb** [comment]
I tried to clean this up, all API changes are now in a separate commit. The last remaining point is that the tests for custom writer must be moved to the new package.

File: pandoc.cabal [hunk 1]
```diff
@@ -560,6 +538,8 @@ library
                    Text.Pandoc.MediaBag,
                    Text.Pandoc.Error,
                    Text.Pandoc.Filter,
+                   Text.Pandoc.Filter.Environment,
+                   Text.Pandoc.Filter.JSON,
```
Comment (tarleb) [inline pandoc.cabal:542]: Maybe the relevant functions should be exported by `Text.Pandoc.Filter`, so these submodules can stay internal?

Comment (jgm) [reply pandoc.cabal:542]: Yes, I think that makes sense.

**Comment by tarleb** [comment]
Good to go, I think.

**Comment by jgm** [comment]
Are you still seeing the large difference in binary size with +lua and -lua?

**Comment by jgm** [comment]
PS. If the gridtables package is ready to release, it would be good to depend on a released version rather than git.

**Comment by tarleb** [comment]
I found that the difference is significant only when linking the Lua library dynamically: there's still a large difference in that case. Compiling a static binary yields a mere 2MB difference.

**Comment by jgm** [comment]
That's surprising that the difference would be *more* with dynamic linking. You'd expect the opposite, wouldn't you?

By the way, I noticed that the command you implemented is `pandoc serve` rather than `pandoc server`.
Was there a reason for this?  I'm not sure, myself, which is better.  But `lua` is a noun, so perhaps that favors `server`.

**Comment by tarleb** [comment]
Gridtables is updated now.

I used `serve` because of this comment: https://github.com/jgm/pandoc/pull/8322#issuecomment-1258439073. I figured it made sense, but now that I think about it, `server` is probably more consitent?

**Comment by tarleb** [comment]
I changed it to `server`.

**Pull request merged**